Adds SPIRE Envoy JWT and OPA tutorial (#23)
spiffe/spire-tutorials: k8s/envoy-jwt-opa/README.md

# Using SPIFFE JWT IDs with Envoy and Open Policy Agent Authorization

[Open Policy Agent](https://www.openpolicyagent.org/) (OPA) is an open source, general-purpose policy engine. The authorization provided by OPA (AuthZ) can be a good complement to the authentication that SPIRE offers (AuthN).

This tutorial builds on the [SPIRE Envoy-JWT Tutorial](../envoy-jwt/README.md) to demonstrate how to combine SPIRE, Envoy and OPA to perform JWT SVID authentication and request authorization. The changes required to implement request authorization with OPA are shown here as a delta to that tutorial, so you should run, or at least read through, the SPIRE Envoy-JWT tutorial first.

To illustrate request authorization with OPA, we add a new sidecar to the backend service used in the SPIRE Envoy JWT tutorial. The new sidecar acts as a new [external authorization filter](https://www.envoyproxy.io/docs/envoy/v1.14.1/intro/arch_overview/security/ext_authz_filter#arch-overview-ext-authz) for Envoy.

> 

Author: Thanks for the email. I applied some changes to improve the diagram.

## Update Deployments

In order to let OPA authorize or reject requests coming to the `backend` service it is necessary to add OPA as a sidecar to the deployment.
We use the `openpolicyagent/opa:latest-istio` image which extends OPA with a gRPC server that implements the Envoy External Authorization API so OPA can communicate policy decisions with Envoy. The new container is added and configured as follow in [`backend-deployment.yaml`](k8s/backend/backend-deployment.yaml):

```console
- name: opa
  image: openpolicyagent/opa:latest-istio
  imagePullPolicy: IfNotPresent
  ports:
    - name: opa-envoy
      containerPort: 8182
      protocol: TCP
    - name: opa-api-port
      containerPort: 8181
      protocol: TCP
   args:
     - "run"
     - "--server"
     - "--config-file=/run/opa/opa-config.yaml"
     - "/run/opa/opa-policy.rego"
   volumeMounts:
     - name: backend-opa-policy
       mountPath: /run/opa
       readOnly: true
```

The ConfigMap `backend-opa-policy` needs to be added into the `volumeMounts` section, like this:

```console
- name: backend-opa-policy
  configMap:
    name: backend-opa-policy-config
```

> The ConfigMap `backend-opa-policy` provides two resources, `opa-config.yaml` described in [OPA Configuration](#opa-configuration) and the `opa-policy.rego` policy explained in the [Rego Policy](#opa-policy) section.

Reviewer: If you think the section should be named Rego Policy instead of OPA Policy, feel free to change the section name on line 107 instead of these cross references.

```suggestion
The ConfigMap `backend-opa-policy` provides two resources, `opa-config.yaml` described in [OPA Configuration](#opa-configuration) and the `opa-policy.rego` policy explained in the [OPA Policy](#opa-policy) section.
```

   [_, payload, _] := io.jwt.decode(encoded_token)
}
```

Consequently, the policy will evaluate to true only when the request is sent to a valid resource (/balances/, /profiles/ or /transactions/) with a `GET` method and the request comes from a workload authenticated with the SPIFFE ID equal to `spiffe://example.org/ns/default/sa/default/frontend`. In all other cases, the request is not authorized by OPA and so is rejected by Envoy.

## Add a New External Authorization Filter to Envoy

Envoy needs to know how to contact the OPA Agent just configured to perform the authorization of every request. To complete the setup, we add a new filter of type External Authorization Filter to the `http_filters` section of the [Envoy configuration](k8s/backend/config/envoy.yaml) as shown below:

```console
- name: envoy.filters.http.ext_authz
  typed_config:
    "@type": type.googleapis.com/envoy.config.filter.http.ext_authz.v2.ExtAuthz
    with_request_body:
      max_request_bytes: 8192
      allow_partial_message: true
   failure_mode_allow: false
   grpc_service:
      google_grpc:
        target_uri: 127.0.0.1:8182
        stat_prefix: ext_authz
      timeout: 0.5s
```

> The configuration tells Envoy to contact the OPA Agent at 127.0.0.1 on port 8182. This matches the configuration for OPA explained at the [OPA Configuration](#opa-configuration) section.

Reviewer: ```suggestion
The configuration tells Envoy to contact the OPA Agent at 127.0.0.1 on port 8182. This matches the configuration for OPA explained in the [OPA Configuration](#opa-configuration) section.
```

## Apply the New Resources

Ensure that the current working directory is `.../spire-tutorials/k8s/envoy-jwt-opa` and deploy the new resources using:

```console
$ kubectl apply -k k8s/.

configmap/backend-envoy configured
configmap/backend-opa-policy-config created
deployment.apps/backend configured
```

For the new configurations to take effect, the `backend` service need to be restarted. Run the following two commands to force the restart:

```console

> kubectl scale deployment backend --replicas=0
> kubectl scale deployment backend --replicas=1

Reviewer: 🤑 

```suggestion
$ kubectl scale deployment backend --replicas=0
$ kubectl scale deployment backend --replicas=1
```

            "content-length": "0",
            "user-agent": "Go-http-client/1.1",
            "x-forwarded-proto": "http",
            "x-request-id": "fad45df6-3cc1-4ce9-9cad-fb3b65eff037"
          },
          "host": "localhost:3001",
          "id": "10476077497628160603",
          "method": "GET",
          "path": "/transactions/1",
          "protocol": "HTTP/1.1"
        },
      ...
      },
      ...
    },
    ...
  },
  ...
  },

  "msg": "Decision Log",
  "query": "data.envoy.authz.allow",
  "requested_by": "",
  "result": true,
  "time": "2020-07-08T20:17:27Z",
  "timestamp": "2020-07-08T20:17:27.7568234Z",
  "type": "openpolicyagent.org/decision_logs"
}
```

> Note the presence of the `authorization` header containing the JWT. As explained in the [Rego Policy](#opa-policy) section, this JWT is decoded using the special-purpose code provided by OPA for dealing with JWTs and then the SPIFFE ID is extracted. As we already know, the SPIFFE ID for the `frontend` service matches the SPIFFE ID defined in the rego policy configured for the OPA Agent. Furthermore, request's path and method also match the rule so the `result` for the decision is `true` and the request is allowed to pass through the filter and reach the `backend` service.

Reviewer: ```suggestion
Note the presence of the `authorization` header containing the JWT. As explained in the [OPA Policy](#opa-policy) section, this JWT is decoded using the special-purpose code provided by OPA for dealing with JWTs and then the SPIFFE ID is extracted. As we already know, the SPIFFE ID for the `frontend` service matches the SPIFFE ID defined in the Rego policy configured for the OPA Agent. Furthermore, request's path and method also match the rule so the `result` for the decision is `true` and the request is allowed to pass through the filter and reach the `backend` service.
```

## Testing Invalid Requests

On the other hand, when you connect to the URL for the `frontend-2` service (e.g. `http://35.222.190.182:3002`), the browser only displays the title without any account details. This is because the SPIFFE ID of the `frontend-2` service (`spiffe://example.org/ns/default/sa/default/frontend-2`) does not satisfy the policy for the OPA Agent.

![Frontend-2-no-details][frontend-2-view-no-details]

[frontend-2-view-no-details]: images/frontend-2_view_no_details.png "Frontend-2 view no details account"

After trying to display the `frontend-2` data, you can verify the decision made by OPA using the same `scripts/backend-opa-logs.sh` script as performed in the previous section. A similar log entry is available for the `frontend-2` service but with the `result` equal to `false` due to the SPIFFE ID mismatch.

## Retesting frontend-2 with a New Policy

> Let's update the rego policy to match the SPIFFE ID of the `frontend-2` service and test again. There is a Bash script that you can leverage to complete this task. Once executed, it will open the editor defined by your `KUBE_EDITOR`, or `EDITOR` environment variables, or fall back to 'vi' for Linux or 'notepad' for Windows.

Reviewer: ```suggestion
Let's update the Rego policy to match the SPIFFE ID of the `frontend-2` service and test again. There is a Bash script that you can leverage to complete this task. Once executed, it will open the editor defined by your `KUBE_EDITOR`, or `EDITOR` environment variables, or fall back to `vi` for Linux or Notepad for Windows.
```

```console
$ bash scripts/backend-update-policy.sh
```

With the editor open, look for the following line that specifies the SPIFFE ID to be matched by the rule:

```console
svc_spiffe_id == "spiffe://example.org/ns/default/sa/default/frontend"
```

Update that line to match the SPIFFE ID for the `frontend-2` service:

```console
svc_spiffe_id == "spiffe://example.org/ns/default/sa/default/frontend-2"
```

> Save the changes and exit. The `backend-update-policy.sh` script resumes. The script applies new version of the ConfigMap and the restart the `backend` pod to pick up the new rule.

Reviewer: ```suggestion
Save the changes and exit. The `backend-update-policy.sh` script resumes. The script applies new version of the ConfigMap and then restarts the `backend` pod to pick up the new rule.
```

Wait some seconds for the deployment to propagate before trying to view the `frontend-2` service in your browser again.
Once the pod is ready, refresh the browser using the correct URL for the `frontend-2` service (e.g. `http://35.222.190.182:3002`). As a result, now the page shows the account details for user _Alex Fergus_.

![Frontend-2][frontend-2-view]

[frontend-2-view]: images/frontend-2_view.png "Frontend-2 view"

On the other hand, if you now connect to the URL for the `frontend` service (e.g. `http://35.222.164.221:3000`), the browser only displays the title without any account details. This is expected because the policy was updated and now the SPIFFE ID for the `frontend` service does not satisfy the policy anymore.

# Cleanup

When you are finished, you can use the following commands to clean the environment created for the tutorial. It will remove:
* All resources created for this SPIRE - Envoy JWT with OPA integration tutorial
* All resources created for the SPIRE - Envoy JWT integration tutorial

> * All deployments and configurations for the SPIRE agent, SPIRE server, and namespace

Reviewer: ```suggestion
* All deployments and configurations for the SPIRE Agent, SPIRE Server, and namespace
```
